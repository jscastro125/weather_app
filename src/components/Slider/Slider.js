import React, {useRef, useState, useEffect} from 'react';
import PropTypes from 'prop-types';
import styles from './Slider.module.css'

const Slider = ({ min, max, value, label, onChange }) => {
    const inputEl = useRef(null);
    const [width, setWidth] = useState(0);
    const [xPos, setXPos] = useState(0);

    useEffect(() => {
        if(inputEl && typeof inputEl.current?.getBoundingClientRect === 'function' ) {
            setWidth(inputEl.current.getBoundingClientRect().width);
            setXPos(inputEl.current.getBoundingClientRect().x)
        }
    }, [inputEl])

    const ratio = (width - 20) / 100;
    const newValue = ((value - min) * 100) / (max - min);
    const newPosition = (xPos) + newValue * ratio;

    return (
        <div className={styles.sliderContainer}>
            <div className={styles.sliderLabel}>
                {label}
            </div>
            <input ref={inputEl} className={styles.customSlider} type="range" min={min} max={max} value={value} onChange={(event) => onChange(event.target.value)}/>
            <div className={styles.rangeValue}>
              <span style={{ left: `calc(${newValue}% + (${newPosition}px))` }}>
                {value}
              </span>
            </div>
        </div>
    );
};

Slider.propTypes = {
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    value: PropTypes.number.isRequired,
    label: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired
}

export default Slider;
