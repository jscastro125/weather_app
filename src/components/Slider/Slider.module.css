.customSlider {
    overflow: visible;
    margin: 0 50px 30px;
    background: #000;
    outline: none;
    opacity: 0.7;
    -webkit-transition: .2s;
    transition: opacity .2s;
}

input[type="range"] {
    -webkit-appearance: none;
}

input[type="range"]::-webkit-slider-runnable-track {
    width: 100%;
    height: 4px;
    cursor: pointer;
    background: #767677;
    border-radius: 25px;
}

input[type="range"]::-webkit-slider-thumb {
    height: 20px;
    width: 20px;
    border-radius: 50%;
    border: 2px solid #000;
    background: #fff;
    cursor: pointer;
    -webkit-appearance: none;
    margin-top: -8px;
}

.sliderContainer {
    position: relative;
    box-sizing: border-box;
    margin-top: 10px;
    margin-bottom: 10px;
}

.sliderLabel {
    margin-top: 10px;
    text-align: center;
}

.rangeValue {
    position: absolute;
    top: 80%;
}

.rangeValue span {
    position: absolute;
    font-size: 16px;
}
