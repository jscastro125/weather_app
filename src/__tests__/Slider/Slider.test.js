import React from "react";
import { create } from "react-test-renderer";
import Slider from '../../components/Slider/Slider';

describe("<Slider />", () => {
    it("should match snapshot", () => {
        const component = create(<Slider max={100} min={10} value={20} label="test" onChange={() => {}} />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
});
