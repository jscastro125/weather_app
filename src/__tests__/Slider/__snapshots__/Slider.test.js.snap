// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> should match snapshot 1`] = `
<div
  className="sliderContainer"
>
  <div
    className="sliderLabel"
  >
    test
  </div>
  <input
    className="customSlider"
    max={100}
    min={10}
    onChange={[Function]}
    type="range"
    value={20}
  />
  <div
    className="rangeValue"
  >
    <span
      style={
        Object {
          "left": "calc(11.11111111111111% + (-2.2222222222222223px))",
        }
      }
    >
      20
    </span>
  </div>
</div>
`;
